### PR TITLE
tests: new spread test for physical-memory-control interface

### PR DESCRIPTION
**Pull request by sergiocazzolato**

### CODE REVIEW

File: tests/lib/snaps/test-snapd-physical-memory-control/Makefile [hunk 1]
```diff
@@ -0,0 +1,5 @@
+CFLAGS += -Wall -Werror
+.PHONY: all
+all: physical-mem
+
+physical-mem: physical-mem.c
```
Comment (bboozzoo) [inline tests/lib/snaps/test-snapd-physical-memory-control/Makefile]: Missing newline

File: tests/lib/snaps/test-snapd-physical-memory-control/snapcraft.yaml [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+apps:
+  physical-mem:
+    command: physical-mem
+    plugs: [physical-memory-control]
+
+parts:
+    physical-mem:
+        source: .
+        plugin: make
+        artifacts: [physical-mem]
```
Comment (bboozzoo) [inline tests/lib/snaps/test-snapd-physical-memory-control/snapcraft.yaml]: newline

File: tests/lib/snaps/test-snapd-physical-memory-control/physical-mem.c [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    /* Map one page */
+    map_base = mmap(NULL, MAP_SIZE, PROT_READ | PROT_WRITE, MAP_SHARED, fd, address & ~MAP_MASK);
+    if (map_base == MAP_FAILED) {
+        perror("cannot map physical memory");
+        goto unmap;
+    }
+    printf("Memory mapped at address %p.\n", map_base);
+
+    /* Read memory map */
+    virt_addr = (char*)map_base + (address & MAP_MASK);
```
Comment (bboozzoo) [inline tests/lib/snaps/test-snapd-physical-memory-control/physical-mem.c]: IIRC busybox has a `devmem` applet that does something similar. Maybe you could use it in the snap.

File: tests/main/interfaces-physical-memory-control/task.yaml [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    echo "When the plug is disconnected"
+    snap disconnect test-snapd-physical-memory-control:physical-memory-control
+    snap interfaces | MATCH "$DISCONNECTED_PATTERN"
+
+    echo "Then the snap is not able to access the physical memory"
+    if test-snapd-physical-memory-control.physical-mem 2>${PWD}/call.error; then
+        echo "Expected permission error reading the physical memory with disconnected plug"
+        exit 1
+    fi
+    MATCH "Permission denied" < call.error
```
Comment (bboozzoo) [inline tests/main/interfaces-physical-memory-control/task.yaml]: Newline

File: tests/lib/snaps/test-snapd-physical-memory-control/physical-mem.c [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+#define MAP_MASK (MAP_SIZE - 1)
+
+int main(int argc, char** argv)
+{
+    int fd;
+    void *map_base, *virt_addr;
+    uint32_t read_result, write_result;
+
+    uint32_t writeval = 't';
+    off_t address = 0x00000001;
```
Comment (bboozzoo) [inline tests/lib/snaps/test-snapd-physical-memory-control/physical-mem.c]: I'm not sure that writing to this address is a good idea. Best consult memory map for x86 in this case.

BTW. will this be allowed STRICT_DEVMEM?

Comment (zyga) [reply tests/lib/snaps/test-snapd-physical-memory-control/physical-mem.c]: I second this opinion. What was the expected outcome? How do we expect people to use this interface?

**Pull request closed without merging**